Add timers collector on rosa ci-operator

morenod opened the pull request:
Capture timers on each step of rosa tests where we wait for some command to be completed:

- Preflight tests
- Installation
- Workers Ready
- Machinepools workers Ready
- Upgrade
- Destroy

morenod commented:
/pj-rehearse

jtaleric commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/retest

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-data-path-9nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

svetsa-rh reviewed (changes requested):
@morenod This is a common workflow for many teams and I am not sure if everyone cares about timings for each step nor if they write to their own ES servers (which is a requirement here). 

May be we can consider making this new step optional by introducing a new variable (which defaults to false) and we can set the env to TRUE in all Perf&Scale tests in config files.

svetsa-rh commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-nightly-aws-rosa-non-sts-advanced-f7

svetsa-rh commented:
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.12-amd64-nightly-aws-rosa-sts-f28

venkataanil commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-nightly-x86-control-plane-24nodes periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented on ci-operator/step-registry/rosa/cluster/deprovision/rosa-cluster-deprovision-commands.sh:
```diff
@@ -42,17 +63,33 @@ if [[ -z "$CLUSTER_ID" ]]; then
+    exit 1
+  else
+    echo "Cluster ${CLUSTER_ID} is still alive, waiting 60 seconds for the next check"
+    sleep 60
```
Old code have no ending condition, if there is any failure deleting the cluster, it will never end, I have used the same logic than the provision job to wait up to CLUSTER_TIMEOUT for the cluster to be installed. so here are waiting up to DESTROY_TIMEOUT to cluster to be destroyed

morenod replied:
if we wait for the pod timeout, we cannot manage a failure, it is better to detect that the cluster is not uninstalled during the process, so we have a change to fix/report/log that failure

yasun1 replied:
After we run `rosa delete cluster`, the cluster status will be in 'uninstalling' status, the action can not be rollback. If the cluster fails deprovision, the cluster maybe hung at uninstalling status or fails in error status. The cluster will be kept here, what we can do is debug/report the failure, we can not fix the failure. 
1. 60m is enough for deprovision.
2. There is no difference between your updating and the old code. Your changes only adds some output logs.

mukrishn replied:
while true and check the rosa cluster state inside the loop

morenod replied:
done

morenod commented on ci-operator/step-registry/rosa/cluster/index/rosa-cluster-index-commands.sh:
```diff
@@ -0,0 +1,58 @@
+
+ES_SERVER=$(cat "/secret/host")
+
+if is_valid_json "${SHARED_DIR}/${METADATA_FILE}" && is_valid_es "${ES_SERVER}" "${ES_INDEX}" ; then
```
if there is no ES_SERVER defined, index wont be executed, so any team without ES can execute the job without failures

morenod replied:
did we talked about changing anything here? I dont remember. indexing will be executed only when there will be an ES_SERVER var, which is added only on perfscale workflows, so it following the comment from @yasun1 
@mukrishn @yasun1

mukrishn replied:
Yes, I think we did and fine with implementation with the fact that it will be called only with respective variable is set. @yasun1 please correct us if we are missed to capture anything in this part.

morenod replied:
added a new var, INDEX_ENABLED, set to false by default, and added as true only on openshift-qe-ocp-qe-perfscale-ci-main__rosa jobs

this will be the entry point of the index script, if not set to "true", anything else wont be executed

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -484,9 +487,9 @@ echo "  Enable audit log: ${ENABLE_AUDIT_LOG}"
+if [[ "$ENABLE_AUTOSCALING" == "true" ]]; then
   echo "  Min replicas: ${MIN_REPLICAS}"
-  echo "  Min replicas: ${MAX_REPLICAS}"  
+  echo "  Min replicas: ${MAX_REPLICAS}"
```
changed, good catch

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

svetsa-rh commented:
@yasun1 @xueli181114 could you ptal? Any objection to adding this to common workflow?

yasun1 commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -579,32 +583,45 @@ CLUSTER_ID=$(cat "${CLUSTER_INFO}" | grep '^ID:' | tr -d '[:space:]' | cut -d ':
+        echo "Cluster on ${CLUSTER_STATE} state, waiting 1 second for the next check..."
+        sleep 1
+      fi
+    fi
```
Meaningless to check each status's timer. For provision, what we want is the whole installation time.

morenod replied:
preflight timers are also important, as we have detected regressions on aws checks done before cluster is installed, during the pending and the validating phases

yasun1 replied:
You are using the same item `timers.install` to record the timer of all of the status. At last, if everything goes well, the timers.install is for the 'Ready' status.  You do not record the preflight check timer.

And in this block, the changes are written complex. 
1. First, in a loop, we'd better always put the exit at the top, in an individual level-1 `if`.
2. Always use `if` instead of `if .. else`

morenod replied:
I am not using the same field to record the status, timers.install is a dictionary and each status has its own field

the final status of the timers dictionary will be something like this:

```
  "timers": {
    "install": {
      "status": "ready",
      "waiting": "30",
      "validating": "5",
      "pending": "69",
      "installing": "1935",
      "total_install": "2039"
    },
    "co_wait_time": "0",
    "nodes_ready": "608",
    "destroy": "424",
    "sts_destroy": "18"
  }
```

Inside timers.install, you get a status field with the last status reached by the cluster installation, and a field with the time dedicated to each transition since the cluster is created.

A rosa cluster transition on this chain waiting > validating > pending > installing > ready

mukrishn replied:
Keep the extreme condition(timeout) at the main if block and error state.

morenod replied:
done

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -160,13 +160,16 @@ fi
 
 cat > ${cluster_config_file} << EOF
 {
+  "uuid": "$(uuidgen)",
```
this is the uuid of the json file generated, it is not related to the cluster uuid. an uuid on the document is required to be uploaded to ES

yasun1 replied:
Thanks. How about make the item's name more clearly, e.g. es_uuid?

morenod replied:
I think uuid is better, ES is only the backend, if we want to use another backend, es_uuid makes no sense

yasun1 replied:
You are trying to add a non-existed item of the cluster detail in the cluster-config file, the cluster-config file is also used by the other team. We need to have an obvious naming for the new attribute. Otherwise, as you said, the uuid is a generated randomly, please use the cluster's uuid to upload to ES.

morenod replied:
there is no cluster uuid when this block is executed, this is a unique uuid to identify the execution, it is not related to the cluster. if workload fails, and cluster is not installed, we still need a uuid to identify the execution, there is no reason to link execution uuid to cluster uuid

morenod replied:
removed uuid

smalleni commented:
@yasun1 Thanks for your detailed review. I'd like to explain the purpose and try to find some common ground.
The Perf/Scale team which uses the common workflows available for installing/deprovisioning clusters wants to track the time taken for each step during an install/uninstall to be able to track regressions from release to release. The only way to do this would be to calculate the time at each step of the install and write that to a long term storage like Elasticsearch. The changes introduced by @morenod help us achieve this so we have this concrete data across releases for ROSA. I do understand that this impacts a common workflow but none of the functionality would change as the indexing step is only enabled on defining an ES_SERVER which I would think that jobs that don't care about this wouldn't be using. If we were to make our own workflow, then we would miss out on any improvements/fixes from the common workflow. 

Is there a way we can help capture these important timings to improve our products while causing minimal disruption to others using this workflow?  I'd love to hear your thoughts on this.

yasun1 commented:
Thanks for the information. It is acceptable for me to add the step `rosa-cluster-index` in the common process. Could you mind add more introduction in `rosa-cluster-index-ref.yaml` to improve understanding? @morenod I advise you to add it in the chain `rosa-cluster-deprovision` so all of the workflows will be involved.

morenod commented on ci-operator/step-registry/rosa/cluster/deprovision/rosa-cluster-deprovision-commands.sh:
```diff
@@ -42,17 +63,33 @@ if [[ -z "$CLUSTER_ID" ]]; then
 done
+end_time=$(date +"%s")
+echo "Cluster destroyed after $(( ${end_time} - ${start_time} )) seconds"
+record_cluster "timers" "destroy" $(( "${end_time}" - "${start_time}" ))
```
we record all timers, including the destroy time, to detect regressions

morenod commented on ci-operator/step-registry/rosa/cluster/deprovision/rosa-cluster-deprovision-ref.yaml:
```diff
@@ -15,6 +15,9 @@ ref:
     documentation: If the cluster is deployed with AWS Security Token Service (STS) instead of IAM credentials, it is a STS cluster.
+  - name: DESTROY_TIMEOUT
+    default: "7200"
+    documentation: Set to number of seconds for the cluster to timeout if it's not destroyed.
```
cloned from CLUSTER_TIMEOUT, which is also 7200, we should decrease both

morenod replied:
done

morenod commented on ci-operator/step-registry/rosa/cluster/upgrade/cluster/rosa-cluster-upgrade-cluster-commands.sh:
```diff
@@ -96,7 +117,8 @@ function upgrade_cluster_to () {
     current_version=$(rosa describe cluster -c $cluster_id -o json | jq -r '.openshift_version')
     if [[ "$current_version" == "$recommended_version" ]]; then
-      log "Upgrade the cluster $cluster_id to the openshift version $recommended_version successfully"
+      record_cluster "timers" "ocp_upgrade" $(( $(date +"%s") - "${start_time}" ))
```
if we introduce the version on the name of the field, we wont be able to use it on any visualization tool, like grafana. timers.ocp_upgrade must be unique on the document, it cannot contain dynamic information

yasun1 replied:
We need to fit the multi-path upgrade, e.g. 4.12 -> 4.13 -> 4.14 -> 4.15.

morenod replied:
moved to 
      record_cluster "timers.ocp_upgrade" "${recommended_version}" $(( $(date +"%s") - "${start_time}" ))
each upgrade will have its own field

yasun1 commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -579,32 +583,45 @@ CLUSTER_ID=$(cat "${CLUSTER_INFO}" | grep '^ID:' | tr -d '[:space:]' | cut -d ':
 start_time=$(date +"%s")
+dyn_start_time=${start_time}
+CLUSTER_PREVIOUS_STATE=""
+record_cluster "timers.install" "status" "started"
```
I do not think there is no output for `rosa describe cluster` after the creation command is executed . If you meet, that is a bug.

mukrishn replied:
replace `started` to `empty`

morenod replied:
done

morenod commented on ci-operator/step-registry/rosa/cluster/upgrade/cluster/rosa-cluster-upgrade-cluster-commands.sh:
```diff
@@ -129,7 +151,8 @@ function upgrade_machinepool_to () {
         if [[ "$node_version" =~ ${mp_recommended_version}- ]]; then
-          log "Upgrade the machinepool $mp_id to the openshift version $mp_recommended_version successfully"
+          record_cluster "timers.machinset_upgrade" "${mp_id}" $(( $(date +"%s") - "${start_time}" ))
+          echo "Upgrade the machinepool $mp_id successfully to the openshift version $mp_recommended_version after $(( $(date +"%s") - ${start_time} )) seconds"
```
done

morenod commented:
> Thanks for the information. It is acceptable for me to add the step `rosa-cluster-index` in the common process. Could you mind add more introduction in `rosa-cluster-index-ref.yaml` to improve understanding? @morenod I advise you to add it in the chain `rosa-cluster-deprovision` so all of the workflows will be involved.

added to the deprovision chain and added more info in the description

morenod commented on ci-operator/step-registry/rosa/cluster/wait-ready/operators/rosa-cluster-wait-ready-operators-commands.sh:
```diff
@@ -37,8 +60,8 @@ if [[ "${costatus}" =~ "timed out" ]]; then
   exit 1
 else
-  echo "All cluster operators are done progressing."
+  record_cluster "timers" "co_wait_time" $(( "${end_time}" - "${start_time}" ))
```
if we put the record on the line 50, we dont know if the oc wait clusteroperators has finished because all operators are ok or because of the timeout

setting on the line 65, we ensure that we are saving only valid information when there is no timeout on the log

morenod replied:
@yasun1 can you confirm this is ok?

yasun1 replied:
lgtm

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -579,32 +583,45 @@ CLUSTER_ID=$(cat "${CLUSTER_INFO}" | grep '^ID:' | tr -d '[:space:]' | cut -d ':
-# Watch the hypershift install log
-if [[ "$HOSTED_CP" == "true" ]]; then
-  rosa logs install -c ${CLUSTER_ID} --watch
-fi
```
we cannot watch the installation process, because if we watch for it, command will hold until cluster will be installed to end, and we lose the possibility of getting the timers. 

after the loop, there is a rosa logs install, which is valid for both rosa classic and hosted clusters, so the functionality is the same than setting a watch

morenod replied:
@yasun1 can you confirm deleting this watch is ok?

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -579,32 +583,45 @@ CLUSTER_ID=$(cat "${CLUSTER_INFO}" | grep '^ID:' | tr -d '[:space:]' | cut -d ':
-  sleep 60
   CLUSTER_STATE=$(rosa describe cluster -c "${CLUSTER_ID}" -o json | jq -r '.state')
-  echo "Cluster state: ${CLUSTER_STATE}"
+  current_time=$(date +"%s")
```
restored the echo

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -160,13 +160,16 @@ fi
 cat > ${cluster_config_file} << EOF
 {
+  "uuid": "$(uuidgen)",
+  "timestamp": "$(date -u +"%Y-%m-%dT%H:%M:%SZ")",
```
this is the timestamp of the execution

when executed this step, cluster installation has not even being launched, so there is no rosa describe cluster to be executed

morenod replied:
after talking to @yasun1 we agreed timestamp is ok as it is not related to the cluster, and created before cluster installation is launched, can you confirm?

morenod replied:
removed timestamp

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented on ci-operator/step-registry/rosa/cluster/provision/rosa-cluster-provision-commands.sh:
```diff
@@ -160,13 +160,16 @@ fi
     "channel_group": "${CHANNEL_GROUP}",
-    "raw_id": "${OPENSHIFT_VERSION}"
+    "raw_id": "${OPENSHIFT_VERSION}",
+    "major_version": "$(echo ${OPENSHIFT_VERSION} | awk -F. '{print $1\".\"$2}')"
```
as agreed with @yasun1, having the major_version as field is ok

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

mukrishn reviewed (approved):
lgtm

morenod commented:
@svetsa-rh @yasun1 did you have time to review this?

morenod commented on ci-operator/step-registry/rosa/aws/non-sts/advanced/rosa-aws-non-sts-advanced-workflow.yaml:
```diff
@@ -25,6 +25,7 @@ workflow:
     - ref: rosa-cluster-wait-ready-nodes
     post:
     - chain: rosa-cluster-deprovision
+    - ref: rosa-cluster-index
```
indexing is not related to provision or deprovision, it is an independent step. It is executed after the deprovision step, to collect deprovision timers, but we could execute it at any other point of the process. If we add it as an step inside the deprovision chain, we are linking indexing only to deprovisioning, which is wrong

morenod replied:
moved there

morenod commented:
@yasun1 

added a new var, INDEX_ENABLED, set to false by default, and added as true only on openshift-qe-ocp-qe-perfscale-ci-main__rosa jobs

this will be the entry point of the index script, if not set to "true", anything else wont be executed

svetsa-rh commented on ci-operator/step-registry/rosa/cluster/deprovision/rosa-cluster-deprovision-commands.sh:
```diff
@@ -42,17 +63,42 @@ if [[ -z "$CLUSTER_ID" ]]; then
   rosa delete oidc-provider -c "${CLUSTER_ID}" -y -m auto
+
+  end_time=$(date +"%s")
+  record_cluster "timers" "sts_destroy" $(( "${end_time}" - "${start_time}" ))
```
@morenod Should this be recorded irrespective of STS cluster or not? 
Our current PerfScale tests all create STS clusters only and there is a plan to move to STS by default soon. However, should we ever need to time a non-sts cluster, the record_cluster won't be invoked. Just calling it out so we make informed decision.

morenod replied:
if not using sts, there is no operator-roles or oidc-providers to remove, so no, this section requires the if and the var of STS to be set. if we move, in the future, to only use sts, and we remove the conditional, this block will be a must

svetsa-rh reviewed (commented):
@morenod All changes after introducing INDEX_ENABLED look good to me. 
I will let @yasun1 take a look and comment.

yasun1 commented:
/lgtm

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
@yasun1 @svetsa-rh I have just rebased again and retest, please review and merge if you consider it is ok

yasun1 commented:
@morenod  Sorry for response late. Could you rebase it again?

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.14-ec-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

yasun1 commented:
/retest

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented:
@svetsa-rh yasun1 I have rebased several times.

All tests are passing, can you please merge before I have to rebase again?

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented:
> @morenod Maybe you need to make some adjustments. We made a change to separate wait-ready-cluster to an individual step by [50414](https://github.com/openshift/release/pull/50414).
> 
> I will take a look tomorrow evening to enhance the merge process.

so after this split on two files, we are loosing the more important dataframe just after the cluster installation is executed, where we get information about how many time we dedicate to temporary stages, like waiting, validating or pending. During the time a job ends, and the waiting one starts, cluster is already on "installing" stage. Im going to remove a sleep 60 at the beginning of the wait and test again, so we will get a minute more or data

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

morenod commented:
/retest

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

yasun1 commented on ci-operator/step-registry/rosa/cluster/wait-ready/cluster/rosa-cluster-wait-ready-cluster-commands.sh:
```diff
@@ -118,31 +139,41 @@ log "Waiting for cluster ready..."
 cluster_info_json=$(mktemp)
 start_time=$(date +"%s")
+dyn_start_time=${start_time}
+CLUSTER_PREVIOUS_STATE="init"
```
Won't this be updated to 'claim' too?

morenod commented:
/pj-rehearse periodic-ci-openshift-qe-ocp-qe-perfscale-ci-main-rosa-4.15-nightly-x86-control-plane-24nodes

yasun1 commented:
/lgtm

svetsa-rh commented:
/approve

svetsa-rh commented:
/pj-rehearse ack

morenod commented:
/pj-rehearse ack
